Win32-network/src/System/Win32/NamedPipes.hsc

```diff
@@ -0,0 +1,217 @@
+-- specified for each instance of the pipe.
+--
+-- * 'pIPE_TYPE_BYTE'
+-- * 'pIPE_TYPE_MESSAGE'
+--
+-- One of the following read modes can be specified. Different instances of the
+-- same pipe can specify different read modes.
+--
+-- * 'pIPE_READMODE_BYTE'
+-- * 'pIPE_READMODE_MESSAGE'
+--
+type PipeMode = UINT
+
+#{enum PipeMode,
+ , pIPE_TYPE_BYTE                = PIPE_TYPE_BYTE
+ , pIPE_TYPE_MESSAGE             = PIPE_TYPE_MESSAGE
+ , pIPE_READMODE_BYTE            = PIPE_READMODE_BYTE
+ , pIPE_READMODE_MESSAGE         = PIPE_READMODE_MESSAGE
+ }
+
+-- | If the 'createNamedPipe' @nMaxInstances@ parameter is
+-- 'pIPE_UNLIMITED_INSTANCES', the number of pipe instances that can be created
+-- is limited only by the availability of system resources.
+pIPE_UNLIMITED_INSTANCES :: DWORD
+pIPE_UNLIMITED_INSTANCES = #const PIPE_UNLIMITED_INSTANCES
+
+-- | Creates an instance of a named pipe and returns a handle for subsequent
+-- pipe operations. A named pipe server process uses this function either to
+-- create the first instance of a specific named pipe and establish its basic
+-- attributes or to create a new instance of an existing named pipe.
+--
+-- For full details see
+-- <https://docs.microsoft.com/en-us/windows/desktop/api/winbase/nf-winbase-createnamedpipea>
+--
+createNamedPipe :: String   -- ^ pipe name of form @\.\pipe\{pipename}@
+                -> OpenMode
+                -> PipeMode
+                -> DWORD    -- ^ nMaxInstances
+                -> DWORD    -- ^ nOutBufferSize
+                -> DWORD    -- ^ nInBufferSize
+                -> DWORD    -- ^ nDefaultTimeOut
+                -> Maybe LPSECURITY_ATTRIBUTES
+                -> IO HANDLE
+createNamedPipe name openMode pipeMode
+                nMaxInstances nOutBufferSize nInBufferSize
+                nDefaultTimeOut mb_attr =
+  withTString name $ \ c_name ->
+    failIf (==iNVALID_HANDLE_VALUE) "CreateNamedPipe" $
+      c_CreateNamedPipe c_name openMode pipeMode
+                        nMaxInstances nOutBufferSize nInBufferSize
+                        nDefaultTimeOut (maybePtr mb_attr)
+
+foreign import ccall unsafe "windows.h CreateNamedPipeW"
+  c_CreateNamedPipe :: LPCTSTR
+                    -> DWORD
+                    -> DWORD
+                    -> DWORD
+                    -> DWORD
+                    -> DWORD
+                    -> DWORD
+                    -> LPSECURITY_ATTRIBUTES
+                    -> IO HANDLE
+
+-- | Enables a named pipe server process to wait for a client process to
+-- connect to an instance of a named pipe. A client process connects by
+-- calling the 'createFile' function.
+--
+-- 'connectNamedPipe' can only be called on a fresh pipe, which is not
+-- connected to other client, or after calling @DisconnectNamedPipe@ windows
+-- API (which we don't expose yet).  This is quite different to Berkeley
+-- socket `accept` call.
+--
+connectNamedPipe :: HANDLE -> Maybe LPOVERLAPPED -> IO ()
```

Comment:
This uses the interruptible FFI call, but we could also do the proper async thing with the IOCP.

Indeed according to the docs we must do this, since we will be using pipes opened in overlapped mode:

> If hNamedPipe was opened with FILE_FLAG_OVERLAPPED, the lpOverlapped parameter must not be NULL. It must point to a valid OVERLAPPED structure. If hNamedPipe was opened with FILE_FLAG_OVERLAPPED and lpOverlapped is NULL, the function can incorrectly report that the connect operation is complete.

Reply:
Good point.  For that I exposed `OVERLAPPED` data type in `System.Win32.Types.Overlapped`, I will update the tests which I think are not using it but they should.  `OVERLAPPED` is backported from a recent version of `Win32` package.

Reply:
This makes `connectNamedPipe` asyncronous. We can create a single event and wait on it until the call completes. We will need to test if we can interrupt a thread that's blocked in this way.

Reply:
Yes indeed it would make `connectNamedPipe` asynchronous via IOCP. The blocking for completion should work exactly the same way (or as close as possible) as the IOCP async read and write. No need to use win32 events.

Reply:
Done, works like charm.